src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostOnAutoInsertEndpoint.cs

```diff
@@ -0,0 +1,143 @@
+    private readonly LSPRequestInvoker _requestInvoker = requestInvoker;
+    private readonly ILogger _logger = loggerFactory.GetOrCreateLogger<CohostOnAutoInsertEndpoint>();
+    private readonly IRazorDocumentMappingService _razorDocumentMappingService = razorDocumentMappingService;
+
+    protected override bool MutatesSolutionState => false;
+
+    protected override bool RequiresLSPSolution => true;
+
+    public Registration? GetRegistration(VSInternalClientCapabilities clientCapabilities, DocumentFilter[] filter, RazorCohostRequestContext requestContext)
+    {
+        if (clientCapabilities.SupportsVisualStudioExtensions)
+        {
+            // TODO: Add overload that doesn't use solution
+            //_remoteServiceProvider.TryInvokeAsync()
+            var providerTriggerChars = new string[] { ">" };
+            var singleServerSupport = true; // TODO: always true for now? Long-term plan?
+            return new Registration
+            {
+                Method = VSInternalMethods.OnAutoInsertName,
+                RegisterOptions = new VSInternalDocumentOnAutoInsertOptions()
+                    .EnableOnAutoInsert(singleServerSupport, providerTriggerChars)
+            };
+        }
+
+        return null;
+    }
+
+    protected override RazorTextDocumentIdentifier? GetRazorTextDocumentIdentifier(VSInternalDocumentOnAutoInsertParams request)
+        => request.TextDocument.ToRazorTextDocumentIdentifier();
+
+    protected override async Task<VSInternalDocumentOnAutoInsertResponseItem?> HandleRequestAsync(VSInternalDocumentOnAutoInsertParams request, RazorCohostRequestContext context, CancellationToken cancellationToken)
+    {
+        var razorDocument = context.TextDocument.AssumeNotNull();
+
+        _logger.LogDebug($"Resolving auto-insertion for {razorDocument.FilePath}");
+
+        _logger.LogDebug($"Calling OOP to resolve insertion at {request.Position} invoked by typing '{request.Character}'");
+        var data = await _remoteServiceProvider.TryInvokeAsync<IRemoteAutoInsertService, RemoteInsertTextEdit?>(
+            razorDocument.Project.Solution,
+            (service, solutionInfo, cancellationToken)
+                => service.TryResolveInsertionAsync(
+                        solutionInfo,
+                        razorDocument.Id,
+                        request.Position,
+                        request.Character,
+                        autoCloseTags: true, // TODO: get value from client options
+                        cancellationToken),
+            cancellationToken).ConfigureAwait(false);
+
+        if (data is { } remoteInsertTextEdit)
+        {
+            _logger.LogDebug($"Got insert text edit from OOP {remoteInsertTextEdit}");
+            return RemoteInsertTextEdit.ToLspInsertTextEdit(remoteInsertTextEdit);
+        }
+
+        // If we are here, Razor didn't return anything, so try HTML
```

Comment:
Check out the latest couple of PRs i've done, there is a `RemoteResponse` type that we'd want to use here, so the remote service gets to be very explicit about whether to call Html.

Reply:
Switched to using RemoteResopnse, thanks!